controllers/ciresource_controller_test.go

```diff
@@ -2,248 +2,168 @@ package controllers
+				}),
 		},
 		{
-			name:   "fallback-resource-deletion",
-			cipool: cipool().name("fallback").priority(-1),
-			cir:    cir("0").pool("fallback"),
-
-			onStateChange: map[ofcirv1.CIResourceState]cirStateEvent{
-				ofcirv1.StateAvailable: func(client client.Client, cir *ofcirv1.CIResource) {
-					client.Delete(context.TODO(), cir)
-				},
-			},
-
-			waitUntil: func(cir *ofcirv1.CIResource) bool {
-				return cir == nil
-			},
+			name: "available resource can be moved under maintenance",
+			testCase: newCIResourceScenario().
+				Setup(scenarioPoolWithSingleCir).
+				ReconcileUntil(func(client client.Client, obj *ofcirv1.CIResource) bool {
+					return obj.Status.State == ofcirv1.StateAvailable
+				}).
+				Then(func(t *testing.T, client client.Client, obj *ofcirv1.CIResource) {
+					obj.Spec.State = ofcirv1.StateMaintenance
+					client.Update(context.Background(), obj)
+				}).
+				ReconcileUntil(func(client client.Client, obj *ofcirv1.CIResource) bool {
+					return obj.Status.State == ofcirv1.StateMaintenance
+				}),
+		},
+		{
+			name: "acquire and release an available resource",
+			testCase: newCIResourceScenario().
+				Setup(scenarioPoolWithSingleCir).
+				ReconcileUntil(func(client client.Client, obj *ofcirv1.CIResource) bool {
+					return obj.Status.State == ofcirv1.StateAvailable
+				}).
+				Then(func(t *testing.T, client client.Client, obj *ofcirv1.CIResource) {
+					obj.Spec.State = ofcirv1.StateInUse
+					client.Update(context.Background(), obj)
+				}, "acquire the available resource").
+				ReconcileUntil(func(client client.Client, obj *ofcirv1.CIResource) bool {
+					return obj.Status.State == ofcirv1.StateInUse
+				}),
+		},
+		{
+			name: "released resource gets cleaned and becomes available",
+			testCase: newCIResourceScenario().
+				Setup(scenarioPoolWithSingleCir).
+				ReconcileUntil(func(client client.Client, obj *ofcirv1.CIResource) bool {
+					return obj.Status.State == ofcirv1.StateAvailable
+				}).
+				Then(func(t *testing.T, client client.Client, obj *ofcirv1.CIResource) {
+					obj.Spec.State = ofcirv1.StateInUse
+					client.Update(context.Background(), obj)
+				}, "acquire the available resource").
+				ReconcileUntil(func(client client.Client, obj *ofcirv1.CIResource) bool {
+					return obj.Status.State == ofcirv1.StateInUse
+				}).
+				Then(func(t *testing.T, client client.Client, obj *ofcirv1.CIResource) {
+					obj.Spec.State = ofcirv1.StateAvailable
+					client.Update(context.Background(), obj)
+				}, "release the acquired resource").
+				ReconcileUntil(func(client client.Client, obj *ofcirv1.CIResource) bool {
+					return obj.Status.State == ofcirv1.StateCleaning
+				}).
+				ReconcileUntil(func(client client.Client, obj *ofcirv1.CIResource) bool {
+					return obj.Status.State == ofcirv1.StateCleaningWait
+				}).
+				ReconcileUntil(func(client client.Client, obj *ofcirv1.CIResource) bool {
+					return obj.Status.State == ofcirv1.StateAvailable
+				}),
 		},
```

Comment:
Agree, there could be many other tests, this is just an initial set for kicking-off, but it would be good if we could all followup with additional cases